river/neural_net/mlp.py

```diff
@@ -1,326 +1,416 @@
+    ...         activations=(
+    ...             nn.activations.ReLU,
+    ...             nn.activations.ReLU,
+    ...             nn.activations.ReLU
+    ...         ),
+    ...         optimizer=optim.SGD(1e-4),
+    ...         seed=42
+    ...     )
+    ... )
+
+    >>> dataset = datasets.TrumpApproval()
+    >>> batch_size = 32
+
+    >>> for epoch in range(10):
+    ...     for xb in pd.read_csv(dataset.path, chunksize=batch_size):
+    ...         yb = xb.pop('five_thirty_eight')
+    ...         y_pred = model.predict_many(xb)
+    ...         model = model.learn_many(xb, yb)
+
+    >>> model.predict_many(xb)
+          five_thirty_eight
+    992           39.361609
+    993           46.398536
+    994           42.094086
+    995           40.195802
+    996           40.782954
+    997           40.839678
+    998           40.896403
+    999           48.362659
+    1000          42.021849
+
+    """
+
+    def __init__(
+        self,
+        hidden_dims,
+        activations,
+        loss: optim.losses.Loss = None,
+        optimizer: optim.base.Optimizer = None,
+        seed: int = None,
+    ):
+        super().__init__(
+            hidden_dims=hidden_dims,
+            activations=activations,
+            loss=loss or optim.losses.Squared(),
+            optimizer=optimizer or optim.SGD(0.01),
+            seed=seed,
+        )
+
+    @classmethod
+    def _default_params(self):
+        from . import activations
+
+        return {
+            "hidden_dims": (20,),
+            "activations": (activations.ReLU, activations.ReLU, activations.Identity),
+        }
+
+    def predict_many(self, X):
+        if not hasattr(self, "w"):
+            return pd.DataFrame({0: 0}, index=X.index)
+        return self(X)
+
+    def learn_one(self, x, y):
+        # Multi-output
+        if isinstance(y, dict):
+            return self.learn_many(X=pd.DataFrame([x]), y=pd.DataFrame([y]))
+
+        # Single output
+        return self.learn_many(X=pd.DataFrame([x]), y=pd.Series([y]))
+
+    def predict_one(self, x):
+        y_pred = self.predict_many(X=pd.DataFrame([x]))
+
+        # Multi-output
+        if len(y_pred.columns) > 1:
+            return y_pred.iloc[0].to_dict()
+
+        # Single output
+        return y_pred.iloc[0, 0]
+
+
+
+class MLPBinaryClassifier(base.Classifier, MLP):
+
+    """Multi-layer Perceptron for binary classification.
+
+    This model is still work in progress. Here are some features that still need implementing:
+
+    - `learn_one` and `predict_one` just cast the input `dict` to a single row dataframe and then
+        call `learn_many` and `predict_many` respectively. This is very inefficient.
+    - Not all of the optimizers in the `optim` module can be used as they are not all vectorised.
+    - Emerging and disappearing features are not supported. Each instance/batch has to have the
+        same features.
+    - The gradient haven't been numerically checked.
+

```

Comment:
Could you add an `Examples` section, like `MLPRegressor`?

Reply:
Done